Implements reproduction of Figure 5. RNN Bifurcations

evaisd opened the pull request:
# Module helpers.py with the functions that do the analysis of the RNN.
# Script plot_fig5.py (self explanatory).
# Some Other additions/fixes:
1. Fixed single/double comma issue in train.py. 
2. Fixed the ZRNN to run without the Euler delta term during inference.
3. Added continuation_phase parameter to generate_stimuli that makes it generate null I_stim.
4. Some fixes in the ZRNN code - fixed devices compatibility issues and parameters names.

evaisd committed:
Replaced single comma with double comma in train.py.

evaisd committed:
Changed the ZRNN to run without Euler delta term during inference

evaisd committed:
Added continuation phase (zero I_stim) mode to datasets.generate_stimuli

evaisd committed:
Adapted drive_model_and_save_outputs.py to work with the change in generate_stimuli (more compact code)

evaisd committed:
Changed the names of ZemlianovaRNN inputs to be identical to those in config.yaml

evaisd committed:
Changed EI_mask to be sent to the same device as the model (enables to send the model to different devices after init).

evaisd committed:
Completed helpers.py

evaisd committed:
Finished initial fig5 script

evaisd committed:
Changed plot titles

evaisd committed:
fixed the duration in generate_stimuli

iranroman commented on zrnn/datasets.py:
```diff
@@ -2,15 +2,16 @@
 from torch.utils.data import Dataset
 
-def generate_stimuli(period, T_onset, duration=2.0, dt=0.001, one_dur=0.01):
+def generate_stimuli(period, T_onset, duration=2.0, dt=0.001, one_dur=0.01, continuation_phase: bool = False):
```
I'd say we should change the name of this keyword argument from `continuation_phase` to `I_stim_without_clicks` and flip the boolean logic.

iranroman commented:
Thanks for this awesome work @evaisd . Please also make changes to the README to include the steps necessary to run this analysis.

iranroman commented on zrnn/helpers.py:
```diff
@@ -0,0 +1,258 @@
+#                      bounds: (float, float),
+#                      time_steps_ms: int = 500,
+#                      grid_length: int = 16,
+#                      at_time: int = None) -> dict:
```
if this stuff in comments is not needed, let's delete it please.

evaisd replied:
Fixed in the more recent push.

iranroman commented on zrnn/helpers.py:
can we rename this as `utils.py` please?

iranroman replied:
@evaisd what do you think about this suggestion?

evaisd replied:
I don't mind, will rename it utils.

iranroman commented on zrnn/helpers.py:
```diff
@@ -0,0 +1,258 @@
+from pathlib import Path
+import numpy as np
+import torch
+import pandas as pd
```
please add pandas to `requirements.txt`

evaisd replied:
Added in the more recent push.

iranroman commented:
This looks great overall. I left some minor comments, I might add more whenever I am able to run this to better understand the mechanics.  

Let's definitively add instructions to the README. Right now, for example, I just tried cloning the repo and running `python plot_fig5.py` but got this cryptic error. I'm sure I'm not following some appropriate step, but I have no way of knowing. 

```
ZemlianovaRNN % python plot_fig5.py
Traceback (most recent call last):
  File "/Users/iranroman/models/ZemlianovaRNN/plot_fig5.py", line 135, in <module>
    Fire(main)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lib/python3.12/site-packages/fire/core.py", line 135, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lib/python3.12/site-packages/fire/core.py", line 468, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
                                ^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lib/python3.12/site-packages/fire/core.py", line 684, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/iranroman/models/ZemlianovaRNN/plot_fig5.py", line 130, in main
    _plot_a(span=span, grid_length=grid_length)
  File "/Users/iranroman/models/ZemlianovaRNN/plot_fig5.py", line 44, in _plot_a
    im = _gen_single_plot(data, ax, span, grid_length, center_around_end_point=True)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/iranroman/models/ZemlianovaRNN/plot_fig5.py", line 26, in _gen_single_plot
    end_point = (data['trajectory0'][-1], data['trajectory1'][-1])
                 ~~~~^^^^^^^^^^^^^^^
KeyError: 'trajectory0'
```

iranroman commented:
@kzemlianova please feel free to have a look and leave a review

iranroman commented:
@IamMarcIvanov (i.e. Danish) you too may leave a review

evaisd committed:
- Fixed the plot_fig5.py script so it should run in any environment.
- Added more comprehensive command line structure with argparse.
- Added labels to the plots.

evaisd committed:
- Updated the requirements file.

evaisd committed:
- Replaced "continuation_phase" with "I_stim_without_clicks"

evaisd committed:
- Changed the default value of the frid resolution to be 128.

evaisd committed:
- Updated the readme file after finishing plot_fig5 script.

evaisd commented:
@iranroman I've got the issues fixed. Now it should run smoothly. It failed when you tried executing it because I left a flag I used to avoid repeating some steps during development. I updated the readme as well.

iranroman commented on README_figs/fig_5b.png:
Wow, these figures are truly cool. 

One question: in this figure, is there a specific reason why you used timepoints that are different to those in the paper?

evaisd replied:
It's a combination of me mistaking the initial condition in this plot (in the paper they're not on the balance manifold as I did but on 0), which I have fixed, and the fact that in our simulation the flow seems to be a bit slower. In any case, I have reset the timepoints to be the same as in the paper.

iranroman commented on plot_fig5.py:
can we add a progress bar or some sort of printing statement to let the user know what stage of the analysis we are at?

Also, running this I could not reproduce the qualitative features of the figures you provide. Granted, I was using a set of model weights I already had saved somewhere. Let me re-try from scratch running all pieces of code as you specify in this PR from the top of the README.

iranroman replied:
Nevermind, I could! I may have been using an old set of `best_model.pth` weights.

evaisd replied:
Added progress statements.

iranroman reviewed (changes requested):
Thanks! I left a few more comments. In my view, this is almost there.

evaisd committed:
- Renamed helpers.py -> utils.py

evaisd committed:
- Changed timepoints to be the same as the paper.
- Changed the timepoints plot initial conditions to be at 0.

evaisd committed:
- Changed the default grid res to 128 in all functions.

evaisd committed:
- Added progress messages.

evaisd committed:
- Updated the timepoints plot in the readme figs.

evaisd committed:
- Deleted comment.

iranroman reviewed (approved):
Wow. Truly amazing work @evaisd I'm in awe. We are ready to merge. 

@kzemlianova and @IamMarcIvanov I encourage you to have a look.
